src/Microsoft.Framework.FileSystemGlobbing/Infrastructure/MatcherContext.cs

```diff
@@ -3,275 +3,180 @@
                     {
-                        if (pattern.Test(fileInfo))
-                        {
-                            include = true;
-                            break;
-                        }
-                    }
-                    if (include)
-                    {
-                        foreach (var pattern in ExcludePatternContexts)
-                        {
-                            if (pattern.Test(fileInfo))
-                            {
-                                include = false;
-                                break;
-                            }
-                        }
-                    }
-                    if (include)
-                    {
-                        if (Frame.RelativePath == null)
-                        {
-                            Files.Add(fileInfo.Name);
-                        }
-                        else
-                        {
-                            Files.Add(Frame.RelativePath + "/" + fileInfo.Name);
-                        }
+                        subDirectories.Add(directoryInfo);
                     }
+
                     continue;
                 }
             }
-            Frame.Stage = Stage.Recursing;
+
+            // When there are qualified sub-directories, create a enumerator and attach to 
+            // current frame. The enumerator will be used to walk through the child nodes.
+            if (subDirectories.Count > 0)
+            {
+                top.Children = subDirectories.ToArray();
+                top.NextChildIndex = 0;
+            }
+
+            top.Visited = true;
         }
 
-        private void DoRecursion()
+        private void DoRecursion(FrameData top, Stack<FrameData> stack)
         {
-            if (Frame.Stage != Stage.Recursing)
+            if (!top.Visited)
             {
                 return;
             }
 
-            if (Frame.ActualDirectories == null)
+            if (top.Children == null || top.NextChildIndex >= top.Children.Length)
             {
-                PopFrame();
+                PopFrame(stack);
                 return;
             }
 
-            if (Frame.ActualDirectoryEnumerating == false)
-            {
-                Frame.ActualDirectoryEnumerating = true;
-                Frame.ActualDirectoryEnumerator = Frame.ActualDirectories.GetEnumerator();
-            }
+            var nextChild = top.Children[top.NextChildIndex++];
+
+            PushFrame(new FrameData(nextChild), stack);
+        }
 
-            var moveNext = Frame.ActualDirectoryEnumerator.MoveNext();
-            if (moveNext == false)
+        private void PushFrame(FrameData frame, Stack<FrameData> stack)
+        {
+            if (stack.Count() > 0)
             {
-                PopFrame();
-                return;
+                frame.RelativePath = stack.Peek().RelativePath + '/' + frame.DirectoryInfo.Name;
```

Comment:
Path.Combine does more than just combination, and doesn't use the '/' character, so we'd also need decide if we want to use '/' for the output of this algorithm or let it go back and forth